default/python/AI/EspionageAI.py

```diff
@@ -3,38 +3,46 @@
 import freeOrionAIInterface as fo  # pylint: disable=import-error
 import AIDependencies
 from AIDependencies import ALL_EMPIRES
-from turn_state import state
+from EnumsAI import EmpireMeters
+from freeorion_tools import cache_by_session_with_turnwise_update
 
 
-def get_empire_detection(empire_id):
+def default_empire_detection_strength():
     # TODO doublecheck typical AI research times for Radar, for below default value
-    empire_detection = (AIDependencies.DETECTION_TECH_STRENGTHS["SPY_DETECT_1"]
-                        if fo.currentTurn() < 40 else AIDependencies.DETECTION_TECH_STRENGTHS["SPY_DETECT_2"])
-    empire = None
-    if empire_id != ALL_EMPIRES:
-        empire = fo.getEmpire(empire_id)
+    return (AIDependencies.DETECTION_TECH_STRENGTHS["SPY_DETECT_1"] if fo.currentTurn() < 40
+            else AIDependencies.DETECTION_TECH_STRENGTHS["SPY_DETECT_2"])
+
+
+@cache_by_session_with_turnwise_update
+def get_empire_detection(empire_id):
+    if empire_id == ALL_EMPIRES:
+        return get_max_empire_detection(fo.allEmpireIDs())
+
+    empire = fo.getEmpire(empire_id)
     if empire:
-        # TODO: expose Empire.GetMeter to automatically take into account detection specials for enemies, or search for
-        # them, but the latter can be inaccurate for enemies
-        for techname in sorted(AIDependencies.DETECTION_TECH_STRENGTHS, reverse=True):
-            if empire.techResearched(techname):
-                empire_detection = AIDependencies.DETECTION_TECH_STRENGTHS[techname]
-                break
-        if empire_id == fo.empireID() and state.have_panopticon:
-            empire_detection += AIDependencies.PANOPTICON_DETECTION_BONUS
-    return empire_detection
-
-
-def colony_detectable_by_empire(planet_id, species_name=None, empire_id=ALL_EMPIRES,
-                                future_stealth_bonus=0, default_result=True):
+        return empire.getMeter(EmpireMeters.DETECTION_STRENGTH).initial
+    else:
+        return default_empire_detection_strength()
+
+
+def get_max_empire_detection(empire_list):
+    max_detection = 0
+    for this_empire_id in empire_list:
+        if this_empire_id != fo.empireID():
+            max_detection = max(max_detection, get_empire_detection(this_empire_id))
+    return max_detection
+
+
+def colony_detectable_by_empire(planet_id, species_name=None, empire=ALL_EMPIRES, future_stealth_bonus=0,
+                                default_result=True):
     """
     Predicts if a planet/colony is/will-be detectable by an empire
     :param planet_id: required, the planet of concern
     :type planet_id: int
     :param species_name: will override the existing planet species if provided
     :type species_name: str
-    :param empire_id: empire whose detection ability is of concern
-    :type empire_id: int
+    :param empire: empire (or list of empires) whose detection ability is of concern
```

Comment:
Should probably more precisely state empire_id in the comment as parameter name is now somewhat ambiguous and could refer to an actual empire object.